examples/generative/molecule_generation.py

```diff
@@ -0,0 +1,493 @@
+in the latent space, the decoder network produces a corresponding SMILES string. A
+mutilayer perceptron network estimates the value of target properties associated with
+each molecule.
+
+**Figure (b)** - Gradient-based optimization in continuous latent space. After training a
+surrogate model f(z) to predict the properties of molecules based on their latent
+representation z, we can optimize f(z) with respect to z to find new latent
+representations expected to have high values of desired properties. These new latent
+representations can then be decoded into SMILES strings, at which point their properties
+can be tested empirically.
+"""
+
+"""
+## Dataset
+
+We would be using [**ZINC – A Free Database of Commercially Available Compounds for
+Virtual Screening**](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1360656/) dataset. The
+dataset comes with molecular formula in SMILE representation along with their respective
+molecular properties such as **logP**( Shows the water–octanal partition coefficient ),
+**SAS**( Shows the synthetic accessibility score ) and **QED**( Shows the Qualitative
+Estimate of Drug-likeness ).
+
+"""
+
+"""shell
+!wget
+https://raw.githubusercontent.com/aspuru-guzik-group/chemical_vae/master/models/zinc_properties/250k_rndm_zinc_drugs_clean_3.csv
+"""
+
+"""
+## Setup
+"""
+
+"""shell
+!pip -q install rdkit-pypi
+"""
+
+import ast
+
+import pandas as pd
+import numpy as np
+
+from rdkit import Chem
+from rdkit import RDLogger
+from rdkit.Chem.Draw import IPythonConsole
+from rdkit.Chem.Draw import MolsToGridImage
+
+import tensorflow as tf
+from tensorflow.keras import layers
+
+import matplotlib.pyplot as plt
+
+tf.random.set_seed(2022)
+
+df = pd.read_csv("/content/250k_rndm_zinc_drugs_clean_3.csv")
+df["smiles"] = df["smiles"].apply(lambda s: s.replace("\n", ""))
+df.head()
+
+
+def molecule_from_smiles(smiles):
+    # MolFromSmiles(m, sanitize=True) should be equivalent to
+    # MolFromSmiles(m, sanitize=False) -> SanitizeMol(m) -> AssignStereochemistry(m, ...)
+    molecule = Chem.MolFromSmiles(smiles, sanitize=False)
+
+    # If sanitization is unsuccessful, catch the error, and try again without
+    # the sanitization step that caused the error
+    flag = Chem.SanitizeMol(molecule, catchErrors=True)
+    if flag != Chem.SanitizeFlags.SANITIZE_NONE:
+        Chem.SanitizeMol(molecule, sanitizeOps=Chem.SanitizeFlags.SANITIZE_ALL ^ flag)
+
+    Chem.AssignStereochemistry(molecule, cleanIt=True, force=True)
+    return molecule
+
+
+print(f"SMILES:\t{df.smiles[100]}\nlogP:\t{df.logP[100]}\nqed:\t{df.qed[100]}")
+molecule = molecule_from_smiles(df.iloc[100].smiles)
+print("Molecule:")
+molecule
+
+SMILE_CHARSET = '[" ", "#", ")", "(", "+", "-", "/", "1", "3", "2", "5", "4", "7", \
+                 "6", "=", "@", "C", "B", "F", "I", "H", "O", "N", "S", "[", "]", \
+                 "c", "l", "o", "n", "P", "s", "r"]'
+
+SMILE_CHARSET = ast.literal_eval(SMILE_CHARSET)
```

Comment:
Why is this necessary?

Reply:
To minimize the number of lines of code. I barely managed to find the entire code within 300 lines